prep/FilePrep.cpp

```diff
@@ -0,0 +1,484 @@
+{
+    // The hope is that raw input values are written as the same raw values
+    // on output. This may not be possible if the input files have different
+    // scaling or incompatible offsets.
+    auto calcOffset = [](double minval, double maxval, double scale)
+    {
+        double interval = maxval - minval;
+        double spacings = interval / scale;  // Number of quantized values in our range.
+        double halfspacings = spacings / 2;  // Half of that number.
+        double offset = (int32_t)halfspacings * scale; // Round to an int value and scale down.
+        return std::round(minval + offset);  // Add the base (min) value and round to an integer.
+    };
+
+    m_b.xform.offset.x = calcOffset(m_trueBounds.minx, m_trueBounds.maxx, m_b.xform.scale.x);
+    m_b.xform.offset.y = calcOffset(m_trueBounds.miny, m_trueBounds.maxy, m_b.xform.scale.y);
+    m_b.xform.offset.z = calcOffset(m_trueBounds.minz, m_trueBounds.maxz, m_b.xform.scale.z);
+}
+
+bool FilePrep::determineOffsetFromInfos(const std::vector<FileInfo>& infos)
+{
+    std::vector<double> xOffsets;
+    std::vector<double> yOffsets;
+    std::vector<double> zOffsets;
+
+    for (const auto& fi : infos)
+    {
+        if (std::isnan(fi.xform.offset.x) || std::isnan(fi.xform.offset.y) ||
+            std::isnan(fi.xform.offset.z))
+            return false;
+        xOffsets.push_back(fi.xform.offset.x);
+        yOffsets.push_back(fi.xform.offset.y);
+        zOffsets.push_back(fi.xform.offset.z);
+    }
+
+    std::sort(xOffsets.begin(), xOffsets.end());
+    std::sort(yOffsets.begin(), yOffsets.end());
+    std::sort(zOffsets.begin(), zOffsets.end());
+
+    // Size should never be zero, but the checks here avoids any divide-by-zero possibility.
+    if (xOffsets.size())
+        m_b.xform.offset.x = xOffsets[xOffsets.size() / 2];
+    if (yOffsets.size())
+        m_b.xform.offset.y = yOffsets[yOffsets.size() / 2];
+    if (zOffsets.size())
+        m_b.xform.offset.z = zOffsets[zOffsets.size() / 2];
+    return true;
+}
+
+void FilePrep::determineScale(const std::vector<FileInfo>& infos)
+{
+    for (const auto& fi : infos)
+    {
+        if (std::abs(fi.xform.scale.x) > std::abs(m_b.xform.scale.x))
+            m_b.xform.scale.x = fi.xform.scale.x;
+        if (std::abs(fi.xform.scale.y) > std::abs(m_b.xform.scale.y))
+            m_b.xform.scale.y = fi.xform.scale.y;
+        if (std::abs(fi.xform.scale.z) > std::abs(m_b.xform.scale.z))
+            m_b.xform.scale.z = fi.xform.scale.z;
+    }
+
+    if (m_b.xform.scale.x != 0.0 && m_b.xform.scale.y != 0.0 && m_b.xform.scale.z != 0.0)
+        return;
+
+    auto calcScale = [](double scale, double low, double high)
+    {
+        // 2 billion is a little less than the int limit.  We center the data around 0 with the
+        // offset, so we're applying the scale to half the range of the data.
+        double val = high / 2 - low / 2;
+        double power = std::ceil(std::log10(val / 2000000000.0));
```

Comment:
Lots of zeroes here, maybe use the apostrophe thing as a separator like `ChunkSize` does further below.

Reply:
Ha. That code may predate when we were still using C++11. But yes, I'll fix.